GVFS/GVFS.Common/Maintenance/PackfileMaintenanceStep.cs

```diff
@@ -139,9 +139,36 @@ protected override void PerformMaintenance()
 
                 GitProcess.Result expireResult = this.RunGitCommand((process) => process.MultiPackIndexExpire(this.Context.Enlistment.GitObjectsRoot));
                 List<string> staleIdxFiles = this.CleanStaleIdxFiles(out int numDeletionBlocked);
-                this.GetPackFilesInfo(out int expireCount, out long expireSize, out hasKeep);
+                this.GetPackFilesInfo(out int expireCount, out long expireSize, out hasKeep);
+
+                GitProcess.Result verifyResult1 = this.RunGitCommand((process) => process.VerifyMultiPackIndex(this.Context.Enlistment.GitObjectsRoot));
+                if (verifyResult1.ExitCodeIsFailure)
+                {
+                    EventMetadata errorMetadata = this.CreateEventMetadata();
+                    errorMetadata["MultiPackIndexVerifyOutput"] = verifyResult1.Output;
+                    errorMetadata["MultiPackIndexVerifyErrors"] = verifyResult1.Errors;
+                    string multiPackIndexPath = Path.Combine(this.Context.Enlistment.GitPackRoot, "multi-pack-index");
+                    errorMetadata["TryDeleteFileResult"] = this.Context.FileSystem.TryDeleteFile(multiPackIndexPath);
+                    activity.RelatedError(errorMetadata, "multi-pack-index is corrupt after write. Deleting and rewriting.");
+
+                    this.RunGitCommand((process) => process.WriteMultiPackIndex(this.Context.Enlistment.GitObjectsRoot));
+                }
+
                 GitProcess.Result repackResult = this.RunGitCommand((process) => process.MultiPackIndexRepack(this.Context.Enlistment.GitObjectsRoot, this.batchSize));
-                this.GetPackFilesInfo(out int afterCount, out long afterSize, out hasKeep);
+                this.GetPackFilesInfo(out int afterCount, out long afterSize, out hasKeep);
+
+                GitProcess.Result verifyResult2 = this.RunGitCommand((process) => process.VerifyMultiPackIndex(this.Context.Enlistment.GitObjectsRoot));
+                if (verifyResult1.ExitCodeIsFailure)
+                {

```

Comment:
I think you mean verifyResult2.ExitCodeIsFailure :-).  Since the two failure check blocks are similar should we move them to a function?